mino/minogrpc/session/mod.go

```diff
@@ -124,13 +141,34 @@ func NewSession(
 	return sess
 }
 
-// Listen implements session.Session. It listens for the stream to detect when
-// it closes, which will start the closing procedure.
-func (s *session) Listen(stream PacketStream) {
-	defer s.close()
+// GetNumParents implements session.Session. It returns the number of active
+// parents in the session.
+func (s *session) GetNumParents() int {
+	s.parentsLock.RLock()
+	defer s.parentsLock.RUnlock()
+
+	return len(s.parents)
+}
+
+// Listen implements session.Session. It listens for the stream and returns only
+// when the is has been closed.
+func (s *session) Listen(relay Relay, table router.RoutingTable, ready chan struct{}) {
+	defer func() {
+		s.parentsLock.Lock()
+
+		delete(s.parents, relay.GetDistantAddress())
+
+		s.parentsLock.Unlock()
+	}()
+
+	s.parentsLock.Lock()
+	s.parents[relay.GetDistantAddress()] = parent{relay: relay, table: table}
+	s.parentsLock.Unlock()
+
+	close(ready)
```

Comment:
this breaks the general golang's channel closing principle (don't close a channel from the receiver side), but I guess that's ok since it is a very specific use.

Reply:
Yep, it could be a wait group or something else, but I think it's clean and clear this way.